src/components/cuda/cupti_profiler.c

```diff
@@ -0,0 +1,1576 @@
+                .pPriv = NULL,
+                .pRawMetricsConfig = nvpw_metricsConfigCreateParams.pRawMetricsConfig,
+            };
+            NVPW_CALL(NVPW_RawMetricsConfig_DestroyPtr((NVPW_RawMetricsConfig_Destroy_Params *) &rawMetricsConfigDestroyParams), return PAPI_EMISC);
+
+            snprintf(desc + strlen(desc), PAPI_2MAX_STR_LEN - strlen(desc), " Numpass=%d", passes);
+            if (passes > 1)
+                snprintf(desc + strlen(desc), PAPI_2MAX_STR_LEN - strlen(desc), " (multi-pass not supported)");
+        }
+        papi_free(temp);
+    }
+    strcpy(description, desc);
+fn_exit:
+    return errno;
+}
+
+static void free_all_enumerated_metrics(void)
+{
+    COMPDBG("Entering.\n");
+    int gpu_id, found;
+    NVPW_MetricsContext_Destroy_Params metricsContextDestroyParams;
+    if (avail_events == NULL)
+        return;
+    for (gpu_id=0; gpu_id<num_gpus; gpu_id++) {
+        found = find_same_chipname(gpu_id);
+        if (found > -1) {
+            avail_events[gpu_id].num_metrics = 0;
+            avail_events[gpu_id].nv_metrics = NULL;
+            continue;
+        }
+        if (avail_events[gpu_id].metricsContextCreateParams.pMetricsContext) {
+            metricsContextDestroyParams = (NVPW_MetricsContext_Destroy_Params) {
+                .structSize = NVPW_MetricsContext_Destroy_Params_STRUCT_SIZE,
+                .pPriv = NULL,
+                .pMetricsContext = avail_events[gpu_id].metricsContextCreateParams.pMetricsContext,
+            };
+            NVPW_MetricsContext_DestroyPtr(&metricsContextDestroyParams);
+        }
+        if (avail_events[gpu_id].nv_metrics)
+            free_event_name_list( &(avail_events[gpu_id].nv_metrics) );
+    }
+    papi_free(avail_events);
+    avail_events = NULL;
+}
+
+// CUPTI Profiler component API functions
+int cuptip_init(const char **pdisabled_reason)
+{
+    COMPDBG("Entering.\n");
+    int errno = PAPI_OK;
+
+    errno = load_cupti_perf_sym();
+    errno += load_nvpw_sym();
+    if (errno != PAPI_OK) {
+        *pdisabled_reason = "Unable to load CUDA library functions.";
+        goto fn_fail;
+    }
+    num_gpus = get_device_count();
+    if (num_gpus <= 0) {
+        *pdisabled_reason = "No GPUs found on system.";
+        goto fn_fail;
+    }
+
+    errno = initialize_cupti_profiler_api();
+    errno += initialize_perfworks_api();
+    if (errno != PAPI_OK) {
+        *pdisabled_reason = "Unable to initialize CUPTI profiler libraries.";
+        goto fn_fail;
+    }
+    errno = init_all_metrics();
+    if (errno != PAPI_OK)
+        goto fn_fail;
+    errno = cuInitPtr(0);
+    if (errno != CUDA_SUCCESS) {
+        *pdisabled_reason = "Failed to initialize CUDA driver API.";
+        goto fn_fail;
+    }
+    return PAPI_OK;
+fn_fail:
+    return PAPI_EMISC;
+}
+
+int cuptip_control_create(event_list_t *all_event_names, int event_count, int *evt_ids, void **pctl, void **pcu_ctx)
```

Comment:
Why do you pass event ids as well as event names?

Reply:
`all_event_names` is the component global event table, not only the events being added. The `evt_ids` contains the indices of the entries in the global event table to be added.

Reply:
this is fixed in commit https://github.com/icl-utk-edu/papi/pull/12/commits/eca91a2a7f6be7de66724fafee9eef2016ca8b38